### ISSUE TITLE
 Error: Data import that includes "QA Status"

### ISSUE THREAD

**Issue by tylerstright** [CRITFC/Helpdesk#43]
We recently had an employee importing data he had reconstructed to meet the output of data from CDMS.  Specifically for this issue, it meant he had included both the "QA Status" and "QA Comments" fields.  He had included the values of "Ready for QA" in the "QA Status" field.  

I reproduced this error by importing a single file into an an existing dataset, including the two QA fields: QA Status = 'Ready for QA', QA Comments = null.  After passing through the column mapper, location mapper, and the import preview (0 errors), the import still fails:
![image](https://user-images.githubusercontent.com/42146185/98707723-984b9700-233d-11eb-9d15-2e12fc8bc848.png)

In devtools (F12), I found this:
![image](https://user-images.githubusercontent.com/42146185/98707860-bf09cd80-233d-11eb-9415-62bd8a0817f3.png)

The CDMS log file shows an SQL INSERT statement:
2020-11-10 10:08:58.3658|DEBUG|services.Controllers.CDMSController|INSERT INTO SGS_Redd_Detail (ActivityId,ByUserId,EffDt,RowStatusId,RowId,QAStatusId,Count,Latitude,Longitude,ReddSpecies,WPTComments,WPTName,WPTType) VALUES ( 60738,3207,'11/10/2020 10:08:58 AM',0,1,Ready for QA,0,null,null,'S_CHN',null,'BLANK','New Redd')

I went back into the import file and removed the "Ready for QA" value (the 'QA Status' field itself still present), and the import will work.  Removing the QA fields entirely also works.  My thoughts are the QA Status is actually a system field that is assigned a numeric value which only prints as a text string once it has been related to the QAStatus table in CDMS.  The "Ready for QA" value is actually applied to new imports automatically, but receives the numeric value of 6. 

I went back into the file and in the QA Status column inserted the value of "for".  This broke the import.  I believe that the importer is bringing in this column with it's literal value which is causing the INSERT statement in SQL to fail due to the inclusion of the "for" word, which is an SQL keyword/verb.  I included the word "for" in a non-system field in the import, and this didn't affect the import. 

This behavior seems to be specifically associated with including an SQL keyword/verb within the values of a system field.

**Comment by nowinski** [comment]
> My thoughts are the QA Status is actually a system field that is assigned a numeric value which only prints as a text string once it has been related to the QAStatus table in CDMS. The "Ready for QA" value is actually applied to new imports automatically, but receives the numeric value of 6.

Yes! QAStatusId is a system field in the detail tables and only accepts integer values. CDMS services is not expecting a string value for QAStatus and does not wrap the incoming value in quotes while building the insert statement. This explains why SQL Server returns ...

Msg 156, Level 15, State 1, Line 2
Incorrect syntax near the keyword 'for'.

Wrapping that string value in quotes and running the insert statement manually in SQL Server produces a new error:

Msg 245, Level 16, State 1, Line 1
Conversion failed when converting the varchar value 'Ready for QA' to data type int

Possible workaround (untested)--use integers in the import file QAStatus column

Possible long term fix--map QAStatus strings to appropriate integer values during import

**Comment by nowinski** [comment]
Decided to leave as is for now. It is possible to import a file with a QAStatus column with integer values. Otherwise, we can set the dataset config default QAStatus.

**Comment by nowinski** [comment]
Reopening this ticket. Would like the ability to import QA Row Status text values

**Comment by JamminCRITFC** [comment]
colettec
Known error would. Our users use the number associated with the status. However, would be nice to have it read the text on import. Way around would be to assign a secondary Row QA field within the dataset itself and then assign those QA status that the user would like to have.

Row QA is one I have been trying decouple from the original for a long time. This would reduce the backend management in the qa tables for the administrators

![Image](https://github.com/user-attachments/assets/1db85222-568b-4ef2-a648-d80b24d380aa)

![Image](https://github.com/user-attachments/assets/c35e5626-0e4c-4351-8e2c-1b5501ab9a5d)

Jul 31 2023, 1:31 pm

**Comment by JamminCRITFC** [comment]
colettec
@Jammin @clarkw3 Part of our grant ask is to have Innovate rework the Activity and Row QA status. This would mean denormalizing it so that they become an attribute of the header and detail. This would also then remove the need to maintain the ActivityQA, DatasetQAStatus, DatasetQAStatus1, and QAStatus tables.

With this then the fields would become text instead of numeric and can be easily imported. Additionally, each dataset can then control the Activity and Row QA select list independently.

Feb 7 2024, 8:28 am

**Comment by JamminCRITFC** [comment]
skeita
I am working on this issue and making some progress.

QAStatus names are now mapped to ids, and Excel or csv files can either use QAStatus names or ids in the QA Status column.
QA Comments that were not showing in preview and Query/Export grids are now displayed.
An error message is displayed for each row instead of the false positive message.
However, I would like to better understand how QAStatuses are used.
There are, in the QAStatus table, about 10 records. The joining table that links QAStatus and Datasets tables , where possible values for datasets' QAStatus seem to be coming from, seems to have only 2 values (5: Approved, 6: Ready for QA), at least for the sample database I have. Thus, when testing, if I use other QAStatuses than those 2, the result shows only ids, not the QAStatus names because only those 2 values are among the possible values. Now, if I manually add the missing QAStatus in the joining table, everything works. I am wondering if datasets' QAStatuses are not set somewhere in the UI or those 2 values are actually the ones that are needed for the dataset I am using for my test (Kelt Reconditioning -- Collections)

![Image](https://github.com/user-attachments/assets/530988f2-deb1-47a2-822b-cc9eed5b97a1)

![Image](https://github.com/user-attachments/assets/2441cef1-ea57-4ca9-b773-0d562f2a5ce6)

Jul 25 2024, 6:46 am

**Comment by JamminCRITFC** [comment]
Jammin
I confirmed that the 'DatasetQAStatus' table in the CRITFC database contains only two values: 5 (Approved) and 6 (Ready for QA). In the "Default Activity QA" field under Admin > Dataset > Configuration, these are the only selectable options, with no modification allowed. I couldn’t find a setting for this in the UI or any hard-coded references in the codebase. It seems the QA Status field can only be configured by directly editing the 'DatasetQAStatus' table in the SQL database. @georgeclark1 @colettec, could you please provide some clarification?

Jul 29 2024, 12:39 pm

**Comment by JamminCRITFC** [comment]
Jammin
Resolved in v3.2, currently in testing.

Nov 14 2024, 12:01 pm
